Upnp search

yongpingchen opened the pull request:
#### Brief

Gateway agent advertise itself as an UPnP service in LAN. The address (ip and port number) of REST APIs is describe as value of `Location` header of UPnP service.  Mobile app can search gateway-agent in the LAN, gateway agent will response the service. 

On the UI `GatewayWizardFragment`, user not need to input IP and Port number of gateway agent manually. Click the `SEARCH` button beside IP address, an `Found Gateway Service` will pop up, user select one gateway service from the list, then click `SELECT` button. IP and port will be automatically filled.  
#### References:

https://docs.google.com/document/d/1qW95E-_ZFX5myWcYSOXLkVkK_F0sfrON4J6iaql0q2o/edit#heading=h.gnb4uxntvtii

yongpingchen committed:
basic implementation

yongpingchen committed:
Add permission for multicast

yongpingchen committed:
load gateway servie in gateway wizard

yongpingchen committed:
show more info in the list view

yongpingchen committed:
fix UI

nfukuzaki commented on app/src/main/java/com/kii/gatewaysample/ui/fragments/GatewayServicesListFragment.java:
```diff
@@ -0,0 +1,105 @@
+
+    public static GatewayServicesListFragment newFragment(UPnPService[] foundServices){
+        GatewayServicesListFragment fragment = new GatewayServicesListFragment();
+        fragment.foundServices = foundServices;
```
use  setArguments()
http://y-anz-m.blogspot.jp/2012/04/androidfragment-setarguments.html

nfukuzaki commented on app/src/main/java/com/kii/gatewaysample/ui/fragments/GatewayServicesListFragment.java:
```diff
@@ -0,0 +1,105 @@
+        this.listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
+            @Override
+            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
+                String location = ((TextView) view.findViewById(R.id.gateway_description) ).getText().toString();
```
use ButterKnife instead of findViewById.

```
@Bind(R.id.xxxx)
TextView textLocation; // This filed will be injected by ButterKnife
```

yongpingchen replied:
`gateway_description` is from item of list view, so can not bind with `ButterKnife`

nfukuzaki commented on app/src/main/java/com/kii/gatewaysample/ui/fragments/wizard/GatewayWizardFragment.java:
```diff
@@ -118,5 +130,53 @@ private void validateRequiredField() {
+            }else if (location.startsWith("https://")){
+                location = location.substring(8);
+            }
+            String[] parts = location.split(":");
```
I think the port number is optional (Maybe User can use default port).

yongpingchen replied:
Port number is decided by gateway, and it must be in the `Location`.  So we need to set it after get the response of UPnP/SSDP search.

yongpingchen committed:
Use setArguments

nfukuzaki commented on app/src/main/java/com/kii/gatewaysample/utils/UPnPControlPointPromise.java:
```diff
@@ -0,0 +1,133 @@
+                            // Skip status string
+                            String contentString = new String(data, 12, i);
+                            UPnPService foundService = parseService(contentString);
+                            services.add(foundService);
```
should check null value.

yongpingchen committed:
change service list fragment UI action

yongpingchen committed:
check null

yongpingchen committed:
Improve UI: disable button to be clickable during searching gateway service
Make socket receive data time out so that socket can close when there is not gateway service

yongpingchen commented:
@nfukuzaki Fixed. Please confirm

nfukuzaki commented:
LGTM
